logfire/_internal/exporters/tail_sampling.py

```diff
@@ -0,0 +1,145 @@
+from __future__ import annotations
+
+import random
+from dataclasses import dataclass
+from functools import cached_property
+from threading import Lock
+
+from opentelemetry import context
+from opentelemetry.sdk.trace import ReadableSpan, Span, SpanProcessor
+
+from logfire._internal.constants import (
+    ATTRIBUTES_LOG_LEVEL_NUM_KEY,
+    LEVEL_NUMBERS,
+    ONE_SECOND_IN_NANOSECONDS,
+    LevelName,
+)
+from logfire._internal.exporters.wrapper import WrapperSpanProcessor
+
+
+@dataclass
+class TailSamplingOptions:
+    level: LevelName | None = 'notice'
+    """
+    Include all spans/logs with level greater than or equal to this level.
+    If None, spans are not included based on level.
+    """
+
+    duration: float | None = 1.0
+    """
+    Include all spans/logs with duration greater than this duration in seconds.
+    If None, spans are not included based on duration.
+    """
+
+
+@dataclass
+class TraceBuffer:
+    """Arguments of `on_start` and `on_end` for spans in a single trace."""
+
+    started: list[tuple[Span, context.Context | None]]
+    ended: list[ReadableSpan]
+
+    @cached_property
+    def first_span(self) -> Span:
+        return self.started[0][0]
+
+
+class TailSamplingProcessor(WrapperSpanProcessor):
```

Comment:
I'm really starting to feel like we should stop adding layers / wrapping processors like this. I would prefer to have a single processor that handles everything (sampling, batching, retries, etc.). I feel like it could be optimized more and would be easier to understand. It can still have pluggable bits just more explicit and less abstract.

Reply:
This layer wraps all processors, including the console and user-defined processors.

Reply:
I get that. But I feel like we should just make `LogfireSpanProcessor` which does all of those things in one place. In particular we avoid double buffering.